net/net.go

```diff
@@ -913,8 +914,68 @@ func (n *net) putRecords(ctx context.Context, tid thread.ID, lid peer.ID, recs [
 		}
 	}
 
-	connector, appConnected := n.getConnector(tid)
+	var (
+		connector, appConnected = n.getConnector(tid)
+		identity                = &thread.Libp2pPubKey{}
+		readKey                 *sym.Key
+		validate                bool
+	)
+
+	if appConnected {
+		var err error
+		if readKey, err = n.store.ReadKey(tid); err != nil {
+			return err
+		} else if readKey != nil {
+			validate = true
+		}
+	}
+
 	for _, record := range chain {
+		if validate {
+			block, err := record.Value().GetBlock(ctx, n)
+			if err != nil {
+				return err
+			}
+
+			event, ok := block.(*cbor.Event)
+			if !ok {
+				event, err = cbor.EventFromNode(block)
+				if err != nil {
+					return fmt.Errorf("invalid event: %w", err)
+				}
+			}
+
+			dbody, err := event.GetBody(ctx, n, readKey)
+			if err != nil {
+				return err
+			}
+
+			if err = identity.UnmarshalBinary(record.Value().PubKey()); err != nil {
+				return err
+			}
+
+			if err = connector.ValidateNetRecordBody(ctx, dbody, identity); err != nil {
```

Comment:
Move validation to here, before setting head. If validation fails, we can un-store the internal blocks that were added during chain traversal.

Reply:
IIUC this change is to remove the invalid record but still return error, so the next call to it may succeed, but may still not if there are more records being invalid? 

In a general sense, this logic seems to be a bit unusual. Alternative ways could be either cleaning up all invalid records unconditionally and succeed, or returning a list of invalid records with errors so the caller can decide if it's okay to call some other method to "fix" them, but I could totally miss the context here.